### PR TITLE
enhancement #440: migration to react query v5

### LINKED ISSUE #440: [ENHANCEMENT] Migration to React Query v5
We need to implement the migration from the current version of React Query to version 5. The implementation involves several key steps to ensure a smooth transition, leveraging the latest features and improvements of React Query v5 while addressing any potential issues.

**Key Implementation Steps:**

1. **Upgrade Dependencies**:
   - Update the React Query package to version 5 using the package manager (yarn). Ensure that all related dependencies are also compatible with React Query v5.

2. **Codebase Refactoring**:
   - Refactor the codebase to accommodate the changes in the React Query API. This includes updating query hooks, mutations, and any other React Query-related logic to align with v5's new syntax and features.

3. **Address Breaking Changes**:
   - Identify and modify any code that relies on features deprecated or removed in React Query v5. Implement necessary replacements or adjustments to ensure compatibility.

4. **Leverage New Features**:
   - Integrate new features and improvements introduced in React Query v5. This may include using new hooks, improved caching mechanisms, or enhanced dev tools.

5. **Update Configuration**:
   - Adjust React Query configuration settings as required by v5. This includes updating default options, query clients, and any related configurations.

**Benefits of Implementation:**
- **Enhanced Performance**: Take advantage of performance optimizations and improvements introduced in React Query v5.
- **New Capabilities**: Utilize new features and enhancements to improve functionality and developer experience.
- **Modernized Codebase**: Keep the project up-to-date with the latest library versions, ensuring better support and maintainability.

### PR DESCRIPTION
**Pull request by foxsis23**
- migration to react query v5

closes #440

### CODE REVIEW

File: fictadvisor-web/src/app/(main)/(search-pages)/search-form/SearchForm.tsx [hunk 1]
```diff
@@ -82,24 +82,29 @@ const SearchForm: FC<SearchFormProps> = ({
   const [collapsed, setCollapsed] = useState(false);
 
   const toastError = useToastError();
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/(search-pages)/search-form/SearchForm.tsx:84]: desctructure useToastError();

File: fictadvisor-web/src/app/(main)/(search-pages)/search-form/SearchForm.tsx [hunk 2]
```diff
@@ -82,24 +82,29 @@ const SearchForm: FC<SearchFormProps> = ({
...
-  const { data: cathedraData } = useQuery(
-    'all-cathedra',
-    () => CathedraAPI.getAll(),
-    {
-      staleTime: Infinity,
-    },
-  );
+  if (isError) {
+    if (isAxiosError(error)) {
+      toastError.displayError(error);
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/(search-pages)/search-form/SearchForm.tsx:99]: and you can use displayError without checking for axiosError

File: fictadvisor-web/src/app/(main)/(search-pages)/teachers/[teacherId]/page.tsx [hunk 3]
```diff
@@ -37,9 +37,13 @@ export interface TeacherContext {
...
 }
 
-const PersonalTeacher: FC<PersonalTeacherProps> = ({ params }) => {
+const PersonalTeacher = ({ params }: PersonalTeacherProps) => {
   const teacherId = params.teacherId;
-  const teacher = useQuery(['teacher'], () => TeacherAPI.get(teacherId)).data;
+  const teacher = useQuery({
+    queryKey: ['teacher', teacherId],
+    queryFn: () => TeacherAPI.get(teacherId),
+  }).data;
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/(search-pages)/teachers/[teacherId]/page.tsx:45]: use destructure, not a .[key]

File: fictadvisor-web/src/app/(main)/(search-pages)/teachers/[teacherId]/page.tsx [hunk 4]
```diff
@@ -37,9 +37,13 @@ export interface TeacherContext {
   teacher: TeacherWithContactsResponse;
 }
 
-const PersonalTeacher: FC<PersonalTeacherProps> = ({ params }) => {
+const PersonalTeacher = ({ params }: PersonalTeacherProps) => {
   const teacherId = params.teacherId;
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/(search-pages)/teachers/[teacherId]/page.tsx:41]: destructure

File: fictadvisor-web/src/app/(main)/(search-pages)/teachers/[teacherId]/personal-teacher-tabs/components/complaint-popup/components/ComplaintPopupContent.tsx [hunk 5]
```diff
@@ -14,18 +13,21 @@ import GroupAPI from '@/lib/api/group/GroupAPI';
...
-    onError: error => {
-      if (isAxiosError(error)) {
-        toastError.displayError(error);
-      }
-    },
   });
+
+  if (error) {
+    if (isAxiosError(error)) {
+      toastError.displayError(error);
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/(search-pages)/teachers/[teacherId]/personal-teacher-tabs/components/complaint-popup/components/ComplaintPopupContent.tsx:28]: same as i wrote earlier

File: fictadvisor-web/src/app/(main)/account/components/general-tab/components/contacts-block/ContactsBlock.tsx [hunk 6]
```diff
@@ -18,18 +18,24 @@ import theme from '@/styles/theme';
...
+  const { data } = useQuery({
+    queryKey: ['contacts', user.id],
+    queryFn: () => UserAPI.getContacts(user.id),
+    refetchOnWindowFocus: false,
+  });
   const contacts = data?.contacts || [];
   const handleClick = () => setIsOpened(!isOpened);
 
+  const refetch = async () => {
+    await qc.refetchQueries({ queryKey: ['contacts', user.id] });
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/account/components/general-tab/components/contacts-block/ContactsBlock.tsx:36]: simple refetch

File: fictadvisor-web/src/app/(main)/account/components/general-tab/components/contacts-block/ContactsBlock.tsx [hunk 7]
```diff
@@ -18,18 +18,24 @@ import theme from '@/styles/theme';
...
 
-  const { data, refetch } = useQuery(
-    'contacts',
-    () => UserAPI.getContacts(user.id),
-    { refetchOnWindowFocus: false },
-  );
+  const { data } = useQuery({
+    queryKey: ['contacts', user.id],
+    queryFn: () => UserAPI.getContacts(user.id),
+    refetchOnWindowFocus: false,
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/account/components/general-tab/components/contacts-block/ContactsBlock.tsx:30]: enabled: !!user

File: fictadvisor-web/src/app/(main)/account/components/general-tab/components/contacts-block/components/ContactItem.tsx [hunk 8]
```diff
@@ -23,7 +22,7 @@ const ContactItem: FC<ContactProps> = ({
   const { user } = useAuthentication();
   const handleDeleteClick = async () => {
     await UserAPI.deleteContact(user.id, id);
-    refetchContacts();
+    await refetchContacts();
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/account/components/general-tab/components/contacts-block/components/ContactItem.tsx:25]: there is a queryClient.invalidateQueries(), so you dont need to pass it as a prop

File: fictadvisor-web/src/app/(main)/account/components/group-tab/GroupTab.tsx [hunk 9]
```diff
@@ -30,17 +30,17 @@ import theme from '@/styles/theme';
...
-    () => GroupService.getGroupData(user, order),
-    {
-      retry: false,
-      refetchOnWindowFocus: false,
-    },
-  );
+  const { data, isLoading } = useQuery({
+    queryKey: ['students', user, order],
+    queryFn: () => GroupService.getGroupData(user, order),
+    retry: false,
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/account/components/group-tab/GroupTab.tsx:41]: enabled: !!user

File: fictadvisor-web/src/app/(main)/account/components/group-tab/GroupTab.tsx [hunk 10]
```diff
@@ -49,6 +49,9 @@ const GroupTab: FC = () => {
 
   const isMobile = useMediaQuery(theme.breakpoints.down('desktop'));
 
+  const refetch = async () => {
+    await qc.refetchQueries({ queryKey: ['students', user, order] });
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/account/components/group-tab/GroupTab.tsx:53]: there is a refetch

File: fictadvisor-web/src/app/(main)/account/components/group-tab/components/table/student-table/components/editing-column/EditingColumn.tsx [hunk 11]
```diff
@@ -47,7 +46,7 @@ import * as styles from './EditingColumn.styles';
 interface EditingColumnProps {
   student: StudentsTableItem;
   permissions: PermissionResponse;
-  refetch: QueryObserverBaseResult['refetch'];
+  refetch: () => Promise<void>;
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/account/components/group-tab/components/table/student-table/components/editing-column/EditingColumn.tsx:49]: remove it

File: fictadvisor-web/src/app/(main)/account/components/group-tab/components/table/types/index.ts [hunk 12]
```diff
@@ -18,14 +17,14 @@ export interface StudentsTableItem extends RequestsTableItem {
 
 export interface RequestsTableProps {
   rows: RequestsTableItem[];
-  refetch: QueryObserverBaseResult['refetch'];
+  refetch: () => Promise<void>;
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/account/components/group-tab/components/table/types/index.ts:20]: you can do it with queryClient.invalidateQueries()

**Review by sifes — CHANGES_REQUESTED** [review]

File: fictadvisor-web/src/app/(main)/account/components/selective-tab/SelectiveTab.tsx [hunk 13]
```diff
@@ -12,13 +12,11 @@ import * as styles from './SelectiveTab.styles';
...
-    ['selectiveDisciplines', user.id],
-    () => UserAPI.getSelectiveDisciplinesBySemester(user.id),
-    {
-      refetchOnWindowFocus: false,
-    },
-  );
+  const { data, isLoading, refetch } = useQuery({
+    queryKey: ['selectiveDisciplines', user.id],
+    queryFn: () => UserAPI.getSelectiveDisciplinesBySemester(user.id),
+    refetchOnWindowFocus: false,
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/account/components/selective-tab/SelectiveTab.tsx:18]: enabled: !!user.id

File: fictadvisor-web/src/app/(main)/account/components/selective-tab/components/opened-selective/OpenedSelective.tsx [hunk 14]
```diff
@@ -29,11 +29,11 @@ const OpenedSelective: FC<OpenedSelectiveProps> = ({
...
   const { user } = useAuthentication();
-  const { data } = useQuery(
-    ['openedSelective', user.id, semester, year],
-    () => UserAPI.getSelectiveDisciplines(user.id, { year, semester }),
-    { refetchOnWindowFocus: false },
-  );
+  const { data } = useQuery({
+    queryKey: ['openedSelective', user.id, semester, year],
+    queryFn: () => UserAPI.getSelectiveDisciplines(user.id, { year, semester }),
+    refetchOnWindowFocus: false,
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/account/components/selective-tab/components/opened-selective/OpenedSelective.tsx:35]: enabled: !!user.id

File: fictadvisor-web/src/app/(main)/schedule/schedule-page/calendar-section/CalendarSection.tsx [hunk 15]
```diff
@@ -31,14 +31,13 @@ export const CalendarSection: FC<CalendarSectionProps> = ({ groups }) => {
...
-      retry: false,
-      refetchOnWindowFocus: false,
-    },
-  );
+  const { data } = useQuery({
+    queryKey: [user?.id, permissionValues],
+    queryFn: () =>
+      PermissionService.getPermissionList(user?.id, permissionValues),
+    retry: false,
+    refetchOnWindowFocus: false,
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/schedule/schedule-page/calendar-section/CalendarSection.tsx:39]: !!teacherId && !!subjectId && !!user?.id,

File: fictadvisor-web/src/app/(main)/schedule/schedule-page/schedule-event-edit-section/schedule-info-card/ScheduleInfoCard.tsx [hunk 16]
```diff
@@ -71,14 +71,13 @@ const ScheduleInfoCard: FC<ScheduleInfoCardProps> = ({
...
-    {
-      retry: false,
-      refetchOnWindowFocus: false,
-    },
-  );
+  const { data } = useQuery({
+    queryKey: [user.id, permissionValues],
+    queryFn: () =>
+      PermissionService.getPermissionList(user.id, permissionValues),
+    retry: false,
```
Comment (sifes) [inline fictadvisor-web/src/app/(main)/schedule/schedule-page/schedule-event-edit-section/schedule-info-card/ScheduleInfoCard.tsx:78]: enabled: !!user.id

File: fictadvisor-web/src/app/admin/comments/page.tsx [hunk 17]
```diff
@@ -23,22 +23,24 @@ const Page = () => {
...
-    ['dates', false],
-    async () => await DatesAPI.getDates(false),
-    useQueryAdminOptions,
-  );
+    ...useQueryAdminOptions,
+  });
+
+  const { data: dates, isLoading: isLoadingDates } = useQuery({
+    queryKey: ['dates', false],
+    queryFn: async () => await DatesAPI.getDates(false),
```
Comment (sifes) [inline fictadvisor-web/src/app/admin/comments/page.tsx:41]: remove async () => await

File: fictadvisor-web/src/app/admin/common/constants/index.ts [hunk 18]
```diff
@@ -1,8 +1,10 @@
 export const useQueryAdminOptions = {
   staleTime: 1500,
-  cacheTime: 1500,
+  gcTime: 1500,
   refetchOnMount: false,
   refetchOnReconnect: false,
   refetchOnWindowFocus: false,
-  keepPreviousData: true,
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (sifes) [inline fictadvisor-web/src/app/admin/common/constants/index.ts:7]: why here a eslint disable

File: fictadvisor-web/src/app/admin/students/edit/[studentId]/components/edit-student-inputs/EditStudentInputs.tsx [hunk 19]
```diff
@@ -12,8 +12,10 @@ import GroupAPI from '@/lib/api/group/GroupAPI';
 import * as styles from './EditStudentInputs.styles';
 
 const EditStudentInputs: FC = () => {
-  const { data: dataGroups } = useQuery('groups', () => GroupAPI.getAll(), {
-    keepPreviousData: true,
+  const { data: dataGroups } = useQuery({
+    queryKey: ['groups'],
+    queryFn: async () => await GroupAPI.getAll(),
```
Comment (sifes) [inline fictadvisor-web/src/app/admin/students/edit/[studentId]/components/edit-student-inputs/EditStudentInputs.tsx:17]: remove async () => await

File: fictadvisor-web/src/app/admin/students/page.tsx [hunk 20]
```diff
@@ -21,16 +21,18 @@ const AdminStudentSearchPage = () => {
   const [currPage, setCurrPage] = useState(0);
   const [pageSize, setPageSize] = useState(5);
 
-  const { data, refetch, isLoading } = useQuery(
-    ['students', currPage, pageSize, queryObj],
-    async () =>
+  const { data, isLoading, refetch } = useQuery({
+    queryKey: ['students', currPage, pageSize, queryObj],
+
+    queryFn: async () =>
```
Comment (sifes) [inline fictadvisor-web/src/app/admin/students/page.tsx:27]: remove async () => await

File: fictadvisor-web/src/app/admin/students/search/components/header-student-search/HeaderStudentSearch.tsx [hunk 21]
```diff
@@ -48,11 +48,11 @@ const HeaderStudentSearch: FC<HeaderStudentSearchProps> = ({ onSubmit }) => {
...
   const [values, setValues] = useState(StudentInitialValues);
 
-  const { data: groups, isLoading } = useQuery(
-    ['groups'],
-    async () => await GroupAPI.getAll(),
-    useQueryAdminOptions,
-  );
+  const { data: groups, isLoading } = useQuery({
+    queryKey: ['groups'],
+    queryFn: async () => await GroupAPI.getAll(),
```
Comment (sifes) [inline fictadvisor-web/src/app/admin/students/search/components/header-student-search/HeaderStudentSearch.tsx:53]: search for all project and remove async () => await in all places

File: fictadvisor-web/src/app/admin/users/edit/[userId]/page.tsx [hunk 22]
```diff
@@ -37,19 +37,19 @@ const AdminUserEdit: FC<AdminUserEditProps> = ({ params }) => {
...
     isSuccess,
     isLoading,
-  } = useQuery(
-    ['getUser', params.userId],
-    () => UserAPI.getUser(params.userId),
-    useQueryAdminOptions,
-  );
+  } = useQuery({
+    queryKey: ['getUser', params.userId],
+    queryFn: async () => await UserAPI.getUser(params.userId),
```
Comment (sifes) [inline fictadvisor-web/src/app/admin/users/edit/[userId]/page.tsx:42]: async () => await

File: fictadvisor-web/src/app/admin/users/page.tsx [hunk 23]
```diff
@@ -22,16 +22,16 @@ const Page = () => {
   const [currPage, setCurrPage] = useState(0);
   const [pageSize, setPageSize] = useState(5);
 
-  const { data, refetch, isLoading } = useQuery(
-    ['users', currPage, pageSize, queryObj],
-    async () =>
+  const { data, isLoading, refetch } = useQuery({
+    queryKey: ['users', currPage, pageSize, queryObj],
+    queryFn: async () =>
```
Comment (sifes) [inline fictadvisor-web/src/app/admin/users/page.tsx:27]: async () => await

File: package-lock.json
Comment (sifes) [inline package-lock.json:1]: remove [package-lock.json](https://github.com/fictadvisor/fictadvisor/pull/477/files#diff-053150b640a7ce75eff69d1a22cae7f0f94ad64ce9a855db544dda0929316519)

**Review by sifes — CHANGES_REQUESTED** [review]

File: fictadvisor-web/src/app/admin/common/constants/index.ts (hunk 18 above)
Comment (foxsis23) [reply fictadvisor-web/src/app/admin/common/constants/index.ts:7]: because i cannot find type for this args

Comment (sifes) [reply fictadvisor-web/src/app/admin/common/constants/index.ts:7]: remove that property completely, as we actually dont really need it

**Review by sifes — CHANGES_REQUESTED** [review]

**Review by sifes — APPROVED** [review]

**Pull request merged**